parsl/providers/azure/azure.py

```diff
@@ -0,0 +1,489 @@
+import json
+import logging
+import os
+import time
+from string import Template
+import base64
+
+from parsl.dataflow.error import ConfigurationError
+from parsl.providers.azure.template import template_string
+from parsl.providers.provider_base import ExecutionProvider
+from parsl.providers.error import OptionalModuleMissing
+from parsl.utils import RepresentationMixin
+from parsl.launchers import SingleNodeLauncher
+
+logger = logging.getLogger(__name__)
+
+try:
+    from azure.common.credentials import ServicePrincipalCredentials
+    from azure.mgmt.resource import ResourceManagementClient
+    from azure.mgmt.network import NetworkManagementClient
+    from azure.mgmt.compute import ComputeManagementClient
+    from azure.mgmt.compute.models import DiskCreateOption
+
+    _api_enabled = True
+
+except ImportError:
+    _api_enabled = False
+else:
+    _api_enabled = True
+
+translate_table = {
+    'VM pending': 'PENDING',
+    'VM running': 'RUNNING',
+    'VM deallocated': 'COMPLETED',
+    'VM stopping': 'COMPLETED',  # We shouldn't really see this state
+    'VM stopped': 'COMPLETED',  # We shouldn't really see this state
+}
+
+
+class AzureProvider(ExecutionProvider, RepresentationMixin):
+    """
+    A Provider for using Microsoft Azure Resources
+
+    One of 2 methods are required to authenticate: keyfile, or environment
+    variables. If  keyfile is not set, the following environment
+    variables must be set: `AZURE_CLIENT_ID` (the access key for
+    your azure account),
+    `AZURE_CLIENT_SECRET` (the secret key for your azure account), the
+    `AZURE_TENANT_ID` (the session key for your azure account), and
+    AZURE_SUBSCRIPTION_ID.
+
+    Parameters
+    ----------
+    vm_reference : dict
+        Dictionary describing the parameters of the Azure VM.
+
+        Required structure:
+        {
+          'publisher': VM OS publisher
+          'offer': VM OS offer
+          'sku': VM OS SKU
+          'version': VM OS version
+          'vm_size': VM Size, analogous to instance type for AWS
+          'disk_size_gb': (int) size of VM disk in gb
+          "admin_username": (str) admin username of VM instances,
+          "password": (str) admin password for VM instances
+        }
+    worker_init : str
+        String to append to the Userdata script executed in the cloudinit phase of
+        instance initialization.
+    walltime : str
+        Walltime requested per block in HH:MM:SS.
+    key_file : str
+        Path to json file that contains 'Azure keys'
+    nodes_per_block : int
```

Comment:
remove this param if it is unused?

Reply:
Same question as with walltime param - will this break compatibility in future releases? Should it be changed in the AWS provider as well?

Reply:
The compatibility argument is the same as for walltime, I think: it is better to remove it now.

A difference between this code and the AWS code is that azure isn't in the last parsl release so we shouldn't be expecting users to already have azure constructors in their workflow, but there might be (in theory) people with AWS constructors that set nodes_per_block.

I expect there aren't though and it's probably fine to remove from both. If you're going to modify AWS, do it in a different PR though.